obj: replace ctree with ravl

pbalcer opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/pmdk/2527)
<!-- Reviewable:end -->

pbalcer commented:
I know there are comments missing, but the rest of the stuff is done. I wanted to run the build before I leave the office.

krzycz commented:
Besides missing comments and a few minor issues/questions <img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
The key question is - why this change? Is it better/faster? How much?

---

Reviewed 54 of 54 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[src/libpmemobj/container_ravl.c, line 47 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34EfvLjA-qLwB-hj11:-L34EfvLjA-qLwB-hj12:b-sutl5) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/container_ravl.c#L47)):*
> ```C
> };
> 
> static int
> ```

/* ... */

---

*[src/libpmemobj/ravl.c, line 2 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34ET-kA9AVOuuNhnFj:-L34ET-kA9AVOuuNhnFk:bkv64wz) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/ravl.c#L2)):*
> ```C
> /*
>  * Copyright 2018, Intel Corporation
> ```

missing comments and LOG/ERR macros

---

*[src/libpmemobj/ravl.c, line 229 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34DoWl4cyD94_5gi9D:-L34DoWl4cyD94_5gi9E:bad0zxp) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/ravl.c#L229)):*
> ```C
> static void
> ravl_node_demote(struct ravl_node *n)
> {
> ```

check for 0 before decr? assert? can it happen?

---

*[src/libpmemobj/ravl.c, line 243 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34E4Re2e-uy2IKSh8f:-L34E4Re2e-uy2IKSh8g:bb6errj) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/ravl.c#L243)):*
> ```C
> 	struct ravl_node *s = ravl_node_sibling(n);
> 	if (!(ravl_node_rank_difference(n) == 0 &&
> 		ravl_node_rank_difference_parent(n->parent, s) > 1))
> ```

+tab

---

*[src/libpmemobj/recycler.c, line 66 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34B39un081J2WZN2YR:-L34B39un081J2WZN2YS:b-sutl5) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/recycler.c#L66)):*
> ```C
> };
> 
> static int
> ```

/* ... */

---

*[src/libpmemobj/tx.c, line 1049 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34BaMOSwBoiJMeFLJh:-L34BaMOSwBoiJMeFLJi:b-sutl5) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/tx.c#L1049)):*
> ```C
> }
> 
> static int
> ```

/* ... */

---

*[src/test/obj_ravl/TEST0, line 43 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L349ux5ZWZTbXqiFZgP:-L349ux5ZWZTbXqiFZgQ:b-tl98c5) ([raw file](https://github.com/pmem/pmdk/blob/697bfa9b878410a533563ea5139ca3ac982d60ff/src/test/obj_ravl/TEST0#L43)):*
> ```
> . ../unittest/unittest.sh
> 
> require_test_type medium
> ```

require_fs_type none

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2527#-:-L34F069Xfi2UoG3C6Ct:bhn6e59)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Initially I just wanted a container with faster removes to use in recycler. Why? Because currently whenever we pfree() something, the allocator doesn't fully update the runtime state to reflect that change, but instead it bumps a counter of "things freed in this run", and only when that counter reaches a certain threshold the runtime state is updated and that object we freed will be visible again (we will be able to allocate in again). Thanks to this lazily updated state, pfree is an extremely fast operation. Updating the runtime state properly at the time of deallocation, which requires 2 tree removes and 1 insert, with ctree has proven to be very slow and, what's worse, it throttles scaling of the the entire allocator because locks are taken for relatively long time. I'm hoping that the RAVL tree, which is a balanced binary tree with a very fast removes (they don't require rotations), will allow me to explore changing this algorithm a bit.

But, once I implemented the thing, I decided that the ctree code has become an unintelligible mess with a terrible API, so I simply substituted it for the RAVL - which I made sure to write as cleanly and verbosely as I could - I'll also add comments to make it more understandable.

You might notice that I also rewritten "tx_add_common", which was another code smell in my opinion. I'll separate that into its own commit.

pbalcer commented:
... and it is slightly faster ;) I'll show some benchmarks tomorrow.
This also allows for arbitrary keys in the tree, which means that we no longer have the 16 bit limitation on indexes in the heap layout. Not sure how useful this is right now, but in the future it might allow us to have zones bigger than 16 gigabytes.

pbalcer commented:
Review status: 39 of 57 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/libpmemobj/ravl.c, line 2 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34ET-kA9AVOuuNhnFj:-L38LuHrw6c-p9PRUP31:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/ravl.c#L2)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

missing comments and LOG/ERR macros
</blockquote></details>

Done.

---

*[src/libpmemobj/ravl.c, line 229 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34DoWl4cyD94_5gi9D:-L38LvmmvWckT-6H46wS:b-pam056) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/ravl.c#L229)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

check for 0 before decr? assert? can it happen?
</blockquote></details>

It can't. Only parents can be demoted.  Look at the ravl_balance function, and you will see that demote is only ever called for nodes that are parents of something. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2527)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Review status: 39 of 57 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/libpmemobj/container_ravl.c, line 47 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34EfvLjA-qLwB-hj11:-L38MVW9HmGpQGQ3LLU0:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/container_ravl.c#L47)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

/* ... */
</blockquote></details>

Done.

---

*[src/libpmemobj/ravl.c, line 243 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34E4Re2e-uy2IKSh8f:-L38MSG8xDrSK5UVs-xS:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/ravl.c#L243)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

+tab
</blockquote></details>

Done.

---

*[src/libpmemobj/recycler.c, line 66 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34B39un081J2WZN2YR:-L38MU4H9Y6zVIvlcoNW:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/recycler.c#L66)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

/* ... */
</blockquote></details>

Done.

---

*[src/libpmemobj/tx.c, line 1049 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34BaMOSwBoiJMeFLJh:-L38MUb5HB5T-oT6fP0m:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/tx.c#L1049)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

/* ... */
</blockquote></details>

Done.

---

*[src/test/obj_ravl/TEST0, line 43 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L349ux5ZWZTbXqiFZgP:-L38MV7itIerNKngv_hB:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/697bfa9b878410a533563ea5139ca3ac982d60ff/src/test/obj_ravl/TEST0#L43)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

require_fs_type none
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2527)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 1 of 54 files at r1, 18 of 18 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/libpmemobj/ravl.c, line 229 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34DoWl4cyD94_5gi9D:-L38SHW2-mZGhLoB1KhE:b-ujoqmi) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/ravl.c#L229)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

It can't. Only parents can be demoted.  Look at the ravl_balance function, and you will see that demote is only ever called for nodes that are parents of something. 
</blockquote></details>

then `assert(rank > 0)` would not break anything 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2527)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Review status: 54 of 57 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/libpmemobj/ravl.c, line 229 at r1](https://reviewable.io:443/reviews/pmem/pmdk/2527#-L34DoWl4cyD94_5gi9D:-L38Yvn68nMJK7gVoD7e:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/eef10b7cd9464ceb27a8d521654adf318b21202e/src/libpmemobj/ravl.c#L229)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

then `assert(rank > 0)` would not break anything 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2527)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2527)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
I've ran some benchmarks and the performance difference is negligible. Which is good, it means that the runtime state tracking isn't really taking any relevant amount of time.

krzycz commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2527)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Pls, rebase.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2527#-:-L3Z4U7w5C0XM4hDtoaI:b5y4rlw)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2527#-:-L3cwpULQ1KIpeRq1a1Q:bnfp4nl)*
<!-- Sent from Reviewable.io -->
